File: courses/codes/HW9-CCC-Lv1/HW9.java
import java.util.Stack;
import java.util.Scanner;

public class HW9 {

    public static boolean recipeCompleted(int N, Scanner scan) {
        Stack<Integer> mountain_top = new Stack<Integer>();
        Stack<Integer> branch = new Stack<Integer>();

        for (int j = 0; j < N; ++j) {
            mountain_top.push(scan.nextInt());
        }

        int n = 1;
        while (n < N) {
            // only need to verify cars #1, 2, ... N-1
            if (!mountain_top.empty() && mountain_top.peek() == n) {
                mountain_top.pop();
                ++n;
            } else if (!branch.empty() && branch.peek() == n) {
                branch.pop();
                ++n;
            } else if (!mountain_top.empty()) {
                branch.push(mountain_top.pop());
            } else {
                return false;
            }
        }
        return true;
    }

    public static void main(String args[]) {

        Scanner scan = new Scanner(System.in);
        int T = scan.nextInt();
        for (int i = 0; i < T; ++i) {
            int N = scan.nextInt();
            System.out.println(recipeCompleted(N, scan)? "Y" : "N");
        }
        scan.close();
    }
}

File: courses/codes/cclv1-jr-class/HW9-CCC-Lv1/hw9.py
def recipe_completed(N):
    mountain_top = []
    branch = []
    for j in range(N):
        mountain_top.append(int(input()))

    n = 1
    while n < N:    # only need to verify cars #1, 2, ... N-1
        if len(mountain_top) > 0 and mountain_top[-1] == n:
            del mountain_top[-1]
            n += 1
        elif len(branch) > 0 and branch[-1] == n:
            del branch[-1]
            n += 1
        elif len(mountain_top) > 0:
            branch.append(mountain_top[-1])
            del mountain_top[-1]
        else:
            return False
    return True


T = int(input())
for i in range(T):
    N = int(input())
    print("Y" if recipe_completed(N) else "N")


